src/core/lib/gpr/useful.h

```diff
@@ -19,48 +19,78 @@
 #ifndef GRPC_CORE_LIB_GPR_USEFUL_H
 #define GRPC_CORE_LIB_GPR_USEFUL_H
 
-/** useful macros that don't belong anywhere else */
+#include <grpc/impl/codegen/port_platform.h>
+
+#include <cstddef>
+
+/** useful utilities that don't belong anywhere else */
+
+namespace grpc_core {
+
+template <typename T>
+T Clamp(T val, T min, T max) {
+  if (val < min) return min;
+  if (val > max) return max;
+  return val;
+}
 
-#define GPR_MIN(a, b) ((a) < (b) ? (a) : (b))
-#define GPR_MAX(a, b) ((a) > (b) ? (a) : (b))
-#define GPR_CLAMP(a, min, max) ((a) < (min) ? (min) : (a) > (max) ? (max) : (a))
 /** rotl, rotr assume x is unsigned */
-#define GPR_ROTL(x, n) (((x) << (n)) | ((x) >> (sizeof(x) * 8 - (n))))
-#define GPR_ROTR(x, n) (((x) >> (n)) | ((x) << (sizeof(x) * 8 - (n))))
+template <typename T>
+constexpr T RotateLeft(T x, T n) {
+  return ((x << n) | (x >> (sizeof(x) * 8 - n)));
+}
+template <typename T>
+constexpr T RotateRight(T x, T n) {
+  return ((x >> n) | (x << (sizeof(x) * 8 - n)));
+}
 
-#define GPR_ARRAY_SIZE(array) (sizeof(array) / sizeof(*(array)))
+// Set the n-th bit of i
+template <typename T>
+T SetBit(T* i, size_t n) {
+  return *i |= (T(1) << n);
+}
 
-#define GPR_SWAP(type, a, b) \
-  do {                       \
-    type x = a;              \
-    (a) = b;                 \
-    (b) = x;                 \
-  } while (0)
+// Clear the n-th bit of i
+template <typename T>
+T ClearBit(T* i, size_t n) {
+  return *i &= ~(T(1) << n);
+}
 
-/** Set the \a n-th bit of \a i (a mutable pointer). */
-#define GPR_BITSET(i, n) ((*(i)) |= (1u << (n)))
+// Get the n-th bit of i
+template <typename T>
+bool GetBit(T i, size_t n) {
+  return (i & (T(1) << n)) != 0;
+}
 
-/** Clear the \a n-th bit of \a i (a mutable pointer). */
-#define GPR_BITCLEAR(i, n) ((*(i)) &= ~(1u << (n)))
+namespace useful_detail {
+inline constexpr uint32_t HexdigitBitcount(uint32_t x) {
+  return (x - ((x >> 1) & 0x77777777) - ((x >> 2) & 0x33333333) -
+          ((x >> 3) & 0x11111111));
+}
+}  // namespace useful_detail
 
-/** Get the \a n-th bit of \a i */
-#define GPR_BITGET(i, n) (((i) & (1u << (n))) != 0)
+inline constexpr uint32_t BitCount(uint32_t i) {
+  return (((useful_detail::HexdigitBitcount(i) +
+            (useful_detail::HexdigitBitcount(i) >> 4)) &
+           0x0f0f0f0f) %
+          255);
+}
 
-#define GPR_INTERNAL_HEXDIGIT_BITCOUNT(x)                        \
-  ((x) - (((x) >> 1) & 0x77777777) - (((x) >> 2) & 0x33333333) - \
-   (((x) >> 3) & 0x11111111))
+template <typename T>
+int icmp(const T& a, const T& b) {
```

Comment:
Suggest calling this something like `IntegerCompare()` or `IntegerCmp()`.

Reply:
I don't like `icmp` but `IntegerCompare` is actually worse because it suggests this function compares integers.

This function compares anything that's comparable and returns an integer, which is a different thing.

Reply:
I'm very tempted to go with `QsortCompatibleComparator` and will do so unless someone comes up with a better plan before midday.

Reply:
Good point about it being more than just integers.  Maybe something like `NumericValueCmp()`, since it's comparing the values *as* numbers, regardless of what actual types they are?

I don't feel super strongly about the name.

Reply:
`QsortCompare` it is